### PR TITLE
Withdraw webhook improvements

### PR DESCRIPTION
**Pull request by motorina0**
### Summary
- add `webhook_body` and `webhook_headers` for `withdraw` callbacks
- store the the `webhook` response in `apipayments`.`extra` column

image

### CODE REVIEW

File: lnbits/core/crud.py [hunk 1]
```diff
@@ -451,6 +451,36 @@ async def update_payment_details(
...
+        "SELECT hash, extra from apipayments WHERE hash = ?",
+        (payment_hash,),
+    )
+    if not row:
+        return
+    existing_extra = json.loads(row["extra"] if row["extra"] else "{}")
+    new_extra = {
+        **existing_extra,
+        **extra,
+    }
```
Comment (callebtc) [inline lnbits/core/crud.py:474]: use `Dict.update()`

File: lnbits/extensions/lnurlp/tasks.py [hunk 2]
```diff
@@ -66,10 +67,4 @@ async def mark_webhook_sent(
     payment.extra["wh_message"] = reason_phrase
     payment.extra["wh_response"] = text
 
-    await core_db.execute(
-        """
-        UPDATE apipayments SET extra = ?
-        WHERE hash = ?
-        """,
-        (json.dumps(payment.extra), payment.payment_hash),
-    )
```
Comment (callebtc) [inline lnbits/extensions/lnurlp/tasks.py:75]: Thank you for this.

File: lnbits/extensions/withdraw/lnurl.py [hunk 3]
```diff
@@ -143,18 +148,39 @@ async def api_lnurl_callback(
...
+                        kwargs["headers"] = json.loads(link.webhook_headers)
+
+                    r: httpx.Response = await client.post(link.webhook_url, **kwargs)
+                    await update_payment_extra(
+                        payment_hash,
+                        dict(
+                            {
+                                "wh_success": r.is_success,
+                                "wh_message": r.reason_phrase,
+                                "wh_response": r.text,
```
Comment (callebtc) [inline lnbits/extensions/withdraw/lnurl.py:171]: `r.text` never has to be checked?

File: lnbits/extensions/withdraw/lnurl.py [hunk 4]
```diff
@@ -143,18 +148,39 @@ async def api_lnurl_callback(
...
                     )
                 except Exception as exc:
                     # webhook fails shouldn't cause the lnurlw to fail since invoice is already paid
-                    logger.error("Caught exception when dispatching webhook url:", exc)
+                    logger.error(
+                        "Caught exception when dispatching webhook url: " + str(exc)
+                    )
+                    await update_payment_extra(
+                        payment_hash,
+                        dict({"wh_success": False, "wh_message": str(exc)}),
```
Comment (callebtc) [inline lnbits/extensions/withdraw/lnurl.py:182]: dict of dict? seems odd.

**Comment by callebtc** [comment]
LGTM, please see comments.

**Review by callebtc — APPROVED** [review]

File: lnbits/extensions/withdraw/lnurl.py (hunk 4 above)
Comment (motorina0) [reply lnbits/extensions/withdraw/lnurl.py]: removed redundant `dict`

File: lnbits/extensions/withdraw/lnurl.py (hunk 3 above)
Comment (motorina0) [reply lnbits/extensions/withdraw/lnurl.py]: - looks safe

```py
    @property
    def text(self) -> str:
        if not hasattr(self, "_text"):
            content = self.content
            if not content:
                self._text = ""
            else:
                decoder = TextDecoder(encoding=self.encoding or "utf-8")
                self._text = "".join([decoder.decode(self.content), decoder.flush()])
        return self._text
```

File: lnbits/core/crud.py (hunk 1 above)
Comment (motorina0) [reply lnbits/core/crud.py]: done!

**Pull request merged**